File: Assets/Lua/SluaTest.lua
GameObject = UnityEngine.GameObject

local M = {}
function M:Start()
    print("Start")
    -- self.transform:Find("btn") 会产生一个临时变量，方法执行完下次Lua.GC会自动回收
    --self.btnTran = self.transform:Find("btn")
    self.btn = self.transform:Find("btn"):GetComponent("Button")
    -- self.btnOnClick 缓存临时变量起来，方法执行完下次Lua.GC不会自动回收，只有当self.btnOnClick=nil后才会被回收
    self.btnOnClick = self.btn.onClick
    self.btn.onClick:AddListener(function()
        print("click")
    end)
end

function M:Update()
    if UnityUtil.GetInputKeyDown(UnityEngine.KeyCode.Q) then
        print("Q")
        --self.transform = nil
        self.btn.onClick:RemoveAllListeners()
    end
end

function M:OnDestroy()
    print("OnDestroy")
    --self.gameObject = nil
    --self.transform = nil
    --self.btn.onClick:RemoveAllListeners()
    --self.btn = nil

    for k, v in pairs(self) do
        print(k, v, type(v))

        if type(v) == 'userdata' then
            --Button
            local mt = getmetatable(v)
            if rawget(mt, "onClick") then
                v["onClick"]:RemoveAllListeners()
            end
            self[k] = nil
        end
    end
end

return M

File: Assets/SluaTest.cs
using System.Collections.Generic;
using System.IO;
using SLua;
using UnityEditor;
using UnityEngine;

public class SluaTest : MonoBehaviour
{
    private LuaSvr L;
    public GameObject prefab;
    private readonly List<GameObject> _objects = new List<GameObject>();

    [MenuItem("Tools/Test")]
    public static void Test()
    {
        var txt = File.ReadAllText($"{Application.dataPath}/Lua/SluaTest.lua");
        var asset = new TextAsset(txt);
        var asset1 = AssetDatabase.LoadAssetAtPath($"Assets/Lua/SluaTest.lua", typeof(UnityEngine.Object));
        Debug.Log(asset.text);
        Debug.Log(asset1 as TextAsset);
    }


    // Start is called before the first frame update
    void Start()
    {
        L = new LuaSvr();
        L.init(tick, Complete, LuaSvrFlag.LSF_BASIC | LuaSvrFlag.LSF_EXTLIB);
    }


    void tick(int p)
    {
        Debug.Log($"tick:{p}");
    }

    private void OnDestroy()
    {
        DelayedDestroy();
    }

    private void DelayedDestroy()
    {
        Debug.Log("DelayedDestroy");
        LuaSvr.mainState.Dispose();
        LuaSvr.mainState = null;
    }


    void Complete()
    {
        // l.start("");
        // object o = LuaSvr.mainState.getFunction("foo").call(1, 2, 3);
        // object[] array = (object[])o;
        // for (int n = 0; n < array.Length; n++)
        //     Debug.Log(array[n]);
        //
        // string s = (string)LuaSvr.mainState.getFunction("str").call(new object[0]);
        Debug.Log("SLua init complete.");
    }

    // Update is called once per frame
    void Update()
    {
        if (Input.GetKeyDown(KeyCode.A))
        {
            var canvas = GameObject.Find("Canvas");
            _objects.Add(Instantiate(prefab, canvas.transform));
        }

        if (Input.GetKeyDown(KeyCode.D))
        {
            foreach (var o in _objects)
            {
                Destroy(o);
            }

            _objects.Clear();
        }
    }
}